sql: TestSavepoints failed

sql.TestSavepoints [failed](https://mesolite.cluster.engflow.com/invocations/default/a1e2a066-1f61-4d00-8819-c2ee3a3a29ee?testReportRun=1&testReportShard=9&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [47a00e3ec37edcbb3ce718a94bb57729c07453f8](https://github.com/cockroachdb/cockroach/commits/47a00e3ec37edcbb3ce718a94bb57729c07453f8):


```
--- FAIL: TestSavepoints (8.82s)
=== RUN   TestSavepoints/rollback_after_lock_timeout
    datadriven.go:263: 
        /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/sql_test_/sql_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/testdata/savepoints:250:
          sql
          CREATE TABLE t(x INT PRIMARY KEY)
          ----
          1: CREATE TABLE t(x INT PRIMARY KEY) -- 0 rows
          -- NoTxn       -> NoTxn       #  (none)
    datadriven.go:263: 
        /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/sql_test_/sql_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/testdata/savepoints:256:
          sql conn=conflict
          BEGIN
          INSERT INTO t VALUES (1)
          ----
          1: BEGIN -- 0 rows
          -- NoTxn       -> Open        #.  (none)
          2: INSERT INTO t VALUES (1) -- 1 row
          -- Open        -> Open        ##  (none)
    datadriven.go:263: 
          /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/sql_test_/sql_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/testdata/savepoints:265:
          sql
          SET lock_timeout = '100ms'
          BEGIN
          SAVEPOINT foo
          SELECT * FROM t WHERE x = 1 FOR UPDATE
          ROLLBACK TO SAVEPOINT foo
          SELECT * FROM t WHERE x = 2 FOR UPDATE
          COMMIT
          SET lock_timeout = 0
        output didn't match expected:
          @@ -2,14 +2,14 @@
           -- NoTxn       -> NoTxn       #.......  (none)
           2: BEGIN -- 0 rows
           -- NoTxn       -> Open        ##......  (none)
           3: SAVEPOINT foo -- 0 rows
           -- Open        -> Open        ###.....  foo
          -4: SELECT * FROM t WHERE x = 1 FOR UPDATE -- pq: canceling statement due to lock timeout on row (x)=(1) in t@t_pkey
          --- Open        -> Aborted     XXXXXXXX  foo
          +4: SELECT * FROM t WHERE x = 1 FOR UPDATE -- 0 rows
          +-- Open        -> Open        ####....  foo
           5: ROLLBACK TO SAVEPOINT foo -- 0 rows
          --- Aborted     -> Open        ###.....  foo
          +-- Open        -> Open        ###.....  foo
           6: SELECT * FROM t WHERE x = 2 FOR UPDATE -- 0 rows
           -- Open        -> Open        ###..#..  foo
           7: COMMIT -- 0 rows
           -- Open        -> NoTxn       ###..##.  (none)
           8: SET lock_timeout = 0 -- 0 rows
--- FAIL: TestSavepoints/rollback_after_lock_timeout (0.65s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=9</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSavepoints.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57190

--- a/pkg/sql/conn_executor_savepoints_test.go
+++ b/pkg/sql/conn_executor_savepoints_test.go
@@ -16,6 +16,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/testutils/datapathutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -30,8 +31,14 @@ func TestSavepoints(t *testing.T) {
 	datadriven.Walk(t, datapathutils.TestDataPath(t, "savepoints"), func(t *testing.T, path string) {
 
 		params := base.TestServerArgs{}
-		s, origConn, _ := serverutils.StartServer(t, params)
-		defer s.Stopper().Stop(ctx)
+		srv, origConn, _ := serverutils.StartServer(t, params)
+		defer srv.Stopper().Stop(ctx)
+		s := srv.ApplicationLayer()
+
+		hostRunner := sqlutils.MakeSQLRunner(srv.SystemLayer().SQLConn(t))
+		// If we have buffered writes enabled, we must have the split lock
+		// reliability enabled (which can be tweaked metamorphically).
+		hostRunner.Exec(t, "SET CLUSTER SETTING kv.lock_table.unreplicated_lock_reliability.split.enabled = true")
 
 		if _, err := origConn.Exec(`SET CLUSTER SETTING kv.transaction.write_buffering.max_buffer_size = '2KiB';`); err != nil {
 			t.Fatal(err)
@@ -62,7 +69,7 @@ func TestSavepoints(t *testing.T) {
 					var ok bool
 					sqlConn, ok = sqlConns[connName]
 					if !ok {
-						sqlConn = s.ApplicationLayer().SQLConn(t, serverutils.DBName(params.UseDatabase))
+						sqlConn = s.SQLConn(t, serverutils.DBName(params.UseDatabase))
 						sqlConns[connName] = sqlConn
 					}
 				}

